source/slang/slang-ir-specialize-function-call.cpp

```diff
@@ -363,9 +364,17 @@ struct FunctionParameterSpecializationContext
             // a new callee function based on the original
             // function and the information we gathered.
             //
-            newFunc = generateSpecializedFunc(oldFunc, funcInfo);
+            newFunc = generateSpecializedFunc(oldFunc, funcInfo, callInfo);
             specializedFuncs.add(callInfo.key, newFunc);
         }
+        else
+        {
+            // We don't need to call this function if the specialized function
+            // is not pre-existing, because the we will specialize the function
+            // based on the call site info exactly, so there is no way that the
+            // arguments mismatch the parameters of the function.
+            typeCastForNewArguments(newFunc, oldCall, callInfo);
```

Comment:
Why is this branch needed? If we had a specialization for int and now calling with uint, shouldn't the tryFindSpecialization fail to find the specialized function in the dictionary?

Reply:
The specialization is always correctly generated for the first call site, that is the reason we need a branch here.

For your example here, I assume you mean
```
int a = ...;
func(a);

uint b = ...
func(b);
```

then yes, tryFindSpecialization will fail in the second call site of `func`, and it will generate two different functions, so this function `typeCastForNewArguments` won't be called. And this is the easiest case.

But this function here is trying to resolve a more tricky case:
```
void func(Texture<float> tex);
```

considering this case, it won't generate two different functions for the two call sites as follow
```
int int_index = ...
func(  g_tex[int_index])


uint uint_index
func(  g_tex[uint_index])
```

the reason is that we use the function name and parameter type as the key in the dictionary, so in the 2nd call site, it will find the function in the dictionary.

Reply:
Right, so we need to find a way to factor that `index` parameter's type into the specialization key in `callInfo`.

Reply:
If we can do this, it will also be straightforward to also encode the uniform-ness of the `index` parameter in the same `callInfo`.